File: app/src/main/java/com/ronju/covid_19tracker/Activitys/Fragment/HomeActivity.java
package com.ronju.covid_19tracker.Activitys.Fragment;

import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;

import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ScrollView;
import android.widget.TextView;
import android.widget.Toast;

import com.bumptech.glide.Glide;
import com.ronju.covid_19tracker.Adapter.LatestUpdateAdapter;
import com.ronju.covid_19tracker.DoInBackground.doInBackground;
import com.ronju.covid_19tracker.LoadingDialog;
import com.ronju.covid_19tracker.Model.WorldDataItem;
import com.ronju.covid_19tracker.R;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import static com.ronju.covid_19tracker.DoInBackground.doInBackground.allCountryData;
import static com.ronju.covid_19tracker.DoInBackground.doInBackground.gAffected;
import static com.ronju.covid_19tracker.DoInBackground.doInBackground.gDeath;
import static com.ronju.covid_19tracker.DoInBackground.doInBackground.gNewAffected;
import static com.ronju.covid_19tracker.DoInBackground.doInBackground.gNewDeath;
import static com.ronju.covid_19tracker.DoInBackground.doInBackground.gNewRecovered;
import static com.ronju.covid_19tracker.DoInBackground.doInBackground.gRecovered;
import static com.ronju.covid_19tracker.DoInBackground.doInBackground.responseFlag1;

public class HomeActivity extends Fragment {
    private boolean flag = true;
    private View view = null;
    private TextView globalAffected, globalNewAffected, globalDeath, globalNewDeath, globalRecovered, globalNewRecovered, countryAffected, countryNewAffected, countryName, countryDeath, countryNewDeath, countryNewRecovered, countryRecovered, changeCountry, date1, date2, date3;
    private SharedPreferences sharedPreferences;
    LoadingDialog loadingDialog;
    RecyclerView latestNewsRecycler;
    LatestUpdateAdapter mAdapter;
    ImageView countryFlag;
    ScrollView scrollView;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        view = inflater.inflate(R.layout.activity_home, container, false);
        //showing toolbar
        ((AppCompatActivity)getActivity()).getSupportActionBar().show();



        //loading dialog
        loadingDialog = new LoadingDialog(getContext());
        loadingDialog.setCancelable(false);
        loadingDialog.setTitle("Loading....");
        loadingDialog.show();

        //initialize view
        initView(view);

//        ElasticityHelper.setUpOverScroll(scrollView);


        //shared preference
        sharedPreferences = getActivity().getSharedPreferences("covid-19_shp", Context.MODE_PRIVATE);


        changeCountry.setOnClickListener(v -> {
           getActivity().getSupportFragmentManager().beginTransaction().replace(R.id.fragmentViewer,new CountryViewActivity()).commit();
        });

        service();
        if (!responseFlag1) {
            setValue();

        }
        return view;
    }


    //getting world data
    private void service() {
        if (responseFlag1) {
            doInBackground countryData = new doInBackground(getContext());
            countryData.CountryDataService(new doInBackground.VolleyResponseListener() {
                @Override
                public void onResponse() {
                    responseFlag1 = false;
                    flag = true;
                    setValue();

                }

                @Override
                public void onErrorResponse() {
                    if (flag) {
                        Toast.makeText(getContext(), "Check Your Internet Connection", Toast.LENGTH_SHORT).show();
                        flag = false;
                    }
                    service();
                }
            });
        }
    }


    public void setValue() {

//latest news adapter
            LinearLayoutManager llm = new LinearLayoutManager(getContext());
            llm.setOrientation(LinearLayoutManager.VERTICAL);
            latestNewsRecycler.setLayoutManager(llm);
            latestNewsRecycler.setHasFixedSize(true);
            mAdapter = new LatestUpdateAdapter();
            latestNewsRecycler.setAdapter(mAdapter);

            //global item
            globalAffected.setText(String.valueOf(gAffected));
            globalNewAffected.setText(String.valueOf(gNewAffected));
            globalDeath.setText(String.valueOf(gDeath));
            globalNewDeath.setText(String.valueOf(gNewDeath));
            globalNewRecovered.setText(String.valueOf(gNewRecovered));
            globalRecovered.setText(String.valueOf(gRecovered));






        //selected country item

        String countryId = sharedPreferences.getString("country_id", "50");
        for (WorldDataItem cItem : allCountryData) {
            if (cItem.getID() == Integer.parseInt(countryId)) {
                countryName.setText(cItem.getCountryName());
                countryAffected.setText(String.valueOf(cItem.getTotalCases()));
                countryNewAffected.setText(String.valueOf(cItem.getTodayCases()));
                countryRecovered.setText(String.valueOf(cItem.getTotalRecovered()));
                countryNewRecovered.setText(String.valueOf(cItem.getTodayRecovered()));
                countryDeath.setText(String.valueOf(cItem.getTotalDeaths()));
                countryNewDeath.setText(String.valueOf(cItem.getTodayDeaths()));
                Glide.with(getContext()).load(cItem.getFlagUrl()).into(countryFlag);
                Calendar calendar = setUpdatedTime(cItem.getUpdateTime());

                date1.setText(new SimpleDateFormat("dd MMM, yyyy, hh:mm a").format(calendar.getTime()));
                date2.setText(new SimpleDateFormat("dd MMM, yyyy, hh:mm a").format(calendar.getTime()));
                date3.setText(new SimpleDateFormat("dd MMM, yyyy").format(calendar.getTime()));

                loadingDialog.dismiss();
                break;
            }
        }
    }
    private Calendar setUpdatedTime(long updateTime) {
        Calendar calendar = Calendar.getInstance();
        calendar.setTimeInMillis(updateTime);
        return calendar;

    }


    //initializing the views
    private void initView(View view) {
        //global section
        globalAffected = view.findViewById(R.id.global_affected);
        globalNewAffected = view.findViewById(R.id.global_new_affected);
        globalDeath = view.findViewById(R.id.global_death);
        globalNewDeath = view.findViewById(R.id.global_new_death);
        globalRecovered = view.findViewById(R.id.global_recovered);
        globalNewRecovered = view.findViewById(R.id.global_new_recovered);

        //country section
        countryName = view.findViewById(R.id.country_name);
        countryAffected = view.findViewById(R.id.country_affected);
        countryNewAffected = view.findViewById(R.id.country_new_affected);
        countryDeath = view.findViewById(R.id.country_death);
        countryNewDeath = view.findViewById(R.id.country_new_death);
        countryRecovered = view.findViewById(R.id.country_recovered);
        countryNewRecovered = view.findViewById(R.id.country_new_recovered);
        countryFlag = view.findViewById(R.id.country_flag_img);

        //others
        date1 = view.findViewById(R.id.update_date1);
        date2 = view.findViewById(R.id.update_date2);
        date3 = view.findViewById(R.id.update_date3);
        changeCountry = view.findViewById(R.id.country_chooser);
        latestNewsRecycler = view.findViewById(R.id.latestNewsRecycler);
    }


}

File: app/src/main/java/com/corona/covid_19tracker/Activity/Fragment/HomeActivity.kt
package com.corona.covid_19tracker.Activity.Fragment

import android.view.LayoutInflater
import android.view.ViewGroup
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import com.bumptech.glide.Glide
import com.corona.covid_19tracker.Adapter.LatestUpdateAdapter
import com.corona.covid_19tracker.Api.CountryDataRetrofitHelper
import com.corona.covid_19tracker.Api.CountryDataService
import com.corona.covid_19tracker.Encryption.Encrypter
import com.corona.covid_19tracker.Model.CountryDataModel
import com.corona.covid_19tracker.R
import com.corona.covid_19tracker.Repository.CountryRepository
import com.corona.covid_19tracker.Repository.Response
import com.corona.covid_19tracker.ViewModel.HomeFragmentViewModel
import com.corona.covid_19tracker.ViewModel.HomeFragmentViewModelFactory
import com.corona.covid_19tracker.databinding.ActivityHomeBinding
import com.corona.covid_19tracker.utils.NetworkUtils
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.lang.Exception
import java.text.SimpleDateFormat
import java.util.*

class HomeActivity : Fragment() {
    private var counter = 0
    private val statisticBundle = Bundle()
    private lateinit var binding: ActivityHomeBinding
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        binding = ActivityHomeBinding.inflate(layoutInflater)
        //showing toolbar
        (activity as AppCompatActivity?)!!.supportActionBar!!.show()
        binding.countryChooser.setOnClickListener {
            activity?.supportFragmentManager?.beginTransaction()
                ?.replace(R.id.fragmentViewer, CountryViewActivity(), "country_view_fragment")
                ?.addToBackStack(null)?.commit();
        };


        val countryDataService =
            CountryDataRetrofitHelper.getInstance().create(CountryDataService::class.java)
        val repository = CountryRepository(countryDataService)
        val viewModel =
            ViewModelProvider(requireActivity(), HomeFragmentViewModelFactory(repository)).get(
                HomeFragmentViewModel::class.java
            )
        viewModel.result.observe(requireActivity(), androidx.lifecycle.Observer {
            when (it) {
                is Response.Loading -> {}
                is Response.Success -> {
                    it.data?.let {
                        if (it.size != 0) {
                            setValue(it)
                            CoroutineScope(Dispatchers.IO).launch()
                            {
                                totalValue(it)
                            }
                            setLatestNewsAdapter(it)
                        }
                    }
                }
                is Response.Error -> {
                    viewModel.getData()
                    if (counter == 0)
                        Toast.makeText(context, it.errorMessage, Toast.LENGTH_SHORT).show()
                    counter++
                }
            }
        })



        binding.countryDBoard.setOnClickListener {
            statisticBundle.putInt("value", 1)
            activity?.supportFragmentManager?.beginTransaction()?.replace(
                R.id.fragmentViewer, StatisticsFragment(statisticBundle), "statistics"
            )?.addToBackStack(null)?.commit()
        }
        binding.globalDBoard.setOnClickListener {
            statisticBundle.putInt("value", 0)
            activity?.supportFragmentManager?.beginTransaction()?.replace(
                R.id.fragmentViewer, StatisticsFragment(statisticBundle), "statistics"
            )?.addToBackStack(null)?.commit()
        }
        return binding.root
    }

    private fun setValue(value: CountryDataModel) {
        var countryId: Int = 50
        try {
            countryId = (requireActivity().application as Encrypter).getDefaultCountry()!!
        } catch (e: Exception) {
        }
        (0 until value.size).forEach {
            if (value[it].countryInfo._id == countryId) {
                context?.let { it1 ->
                    Glide.with(it1.applicationContext).load(value[it].countryInfo.flag)
                        .into(binding.countryFlagImg)
                }

                statisticBundle.putInt("CnewCase", value[it].todayCases)
                statisticBundle.putInt("CnewDeath", value[it].todayDeaths)
                statisticBundle.putInt("CnewRecover", value[it].todayRecovered)
                statisticBundle.putInt("Cactive", value[it].active)
                statisticBundle.putInt("Cserious", value[it].critical)

                binding.countryName.text = value[it].country
                binding.countryAffected.text = value[it].cases.toString()
                binding.countryRecovered.text = value[it].recovered.toString()
                binding.countryDeath.text = value[it].deaths.toString()
                binding.countryNewAffected.text = value[it].todayCases.toString()
                binding.countryNewDeath.text = value[it].todayDeaths.toString()
                binding.countryNewRecovered.text = value[it].todayRecovered.toString()
                val calendar = setUpdatedTime(value[it].updated)
                binding.updateDate1.text =
                    SimpleDateFormat("dd MMM, yyyy, hh:mm a").format(calendar.time).toString()
                statisticBundle.putString(
                    "Date",
                    SimpleDateFormat("dd MMM, yyyy, hh:mm a").format(calendar.time).toString()
                )
                binding.updateDate2.text =
                    SimpleDateFormat("dd MMM, yyyy, hh:mm a").format(calendar.time).toString()
                binding.updateDate3.text =
                    SimpleDateFormat("dd MMM, yyyy").format(calendar.time).toString()
                binding.shimmerViewContainer.stopShimmer()
                binding.shimmerViewContainer.visibility = View.INVISIBLE
                binding.homeScrollView.visibility = View.VISIBLE


                return
            }
        }
    }

    private fun totalValue(value: CountryDataModel) {
        var totalNewCase = 0
        var totalNewDeath = 0
        var totalNewRecover = 0
        var totalCase = 0;
        var totalDeath = 0
        var totalRecover = 0
        var critical = 0
        var active = 0
        (0 until value.size).forEach {
            totalNewCase += value[it].todayCases
            totalNewDeath += value[it].todayDeaths
            totalNewRecover += value[it].todayRecovered
            totalCase += value[it].cases
            totalDeath += value[it].deaths
            totalRecover += value[it].recovered
            critical += value[it].critical
            active += value[it].active
        }
        statisticBundle.putInt("GnewCase", totalNewCase)
        statisticBundle.putInt("GnewDeath", totalNewDeath)
        statisticBundle.putInt("GnewRecover", totalNewRecover)
        statisticBundle.putInt("Gactive", active)
        statisticBundle.putInt("Gserious", critical)
        setGlobal(totalNewCase, totalNewDeath, totalNewRecover, totalCase, totalDeath, totalRecover)
    }

    private fun setGlobal(
        totalNewCase: Int,
        totalNewDeath: Int,
        totalNewRecover: Int,
        totalCase: Int,
        totalDeath: Int,
        totalRecover: Int
    ) {
        binding.globalAffected.text = totalCase.toString()
        binding.globalNewAffected.text = totalNewCase.toString()
        binding.globalDeath.text = totalDeath.toString()
        binding.globalNewDeath.text = totalNewDeath.toString()
        binding.globalRecovered.text = totalRecover.toString()
        binding.globalNewRecovered.text = totalNewRecover.toString()
    }

    private fun setLatestNewsAdapter(it: CountryDataModel) {
        binding.latestNewsRecycler.layoutManager =
            LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false)
        binding.latestNewsRecycler.setHasFixedSize(true)
        binding.latestNewsRecycler.adapter = LatestUpdateAdapter(it)
    }

    private fun setUpdatedTime(updateTime: Long): Calendar {
        val calendar = Calendar.getInstance()
        calendar.timeInMillis = updateTime
        return calendar
    }

}